C++/Indentation Rules.tmPreferences

```diff
@@ -6,26 +6,102 @@
 	<key>settings</key>
 	<dict>
 		<key>decreaseIndentPattern</key>
-		<string>(?x)
-		^ (.*\*/)? \s* \} .* $
-		|   ^ \s* (public|private|protected): \s* $
-		|   ^ \s* @(public|private|protected) \s* $
-		</string>
+		<string><![CDATA[(?x)
+			# line beginning with whitespace or block comments
+			^ (.*\*/)? \s*
+			(?:
+			# dedent closing braces
+			  \}
+			# detent `case ... :`
+			| case\b
+			# detent `default:`
+			| default\b
+			# class member access attributes
+			| ( public | private | protected ):
+			# objc/objc++ attributes
+			| @( public | private | protected )\b
+			)
+		]]></string>
+
 		<key>increaseIndentPattern</key>
-		<string>(?x)
-		^ .* \{ [^}"']* $
-		|   ^ \s* (public|private|protected): \s* $
-		|   ^ \s* @(public|private|protected) \s* $
-		</string>
+		<string><![CDATA[(?x)
+			# line beginning with whitespace or block comments
+			^ (.*\*/)? \s*
+			(?:
+			# indent after opening braces (may be followed by whitespace or comments)
+			# but exclude lines such as `extern "C" {`
+			  (?!extern\s*\"[^\"]\") .* \{ (?: \s* /\*.*\*/ )* \s* (?: //.* )? $
+			# indent after `case ... :`
+			| case\b
+			# indent after `default:`
+			| default\b
+			# class member access attributes
+			| ( public | private | protected ):
+			# objc/objc++ attributes
+			| @( public | private | protected )\b
+			)
+		]]></string>
 
 		<key>bracketIndentNextLinePattern</key>
-		<string>(?x)
-		^ \s* \b(if|while|else)\b [^;]* $
-		| ^ \s* \b(for)\b .* $
-		</string>
+		<string><![CDATA[(?x)
+			# line beginning with whitespace or block comments
+			^ (.*\*/)? (?<comment_or_whitespace> (?: \s* (?<block_comment> /\*.*\*/ ) )* \s* )
```

Comment:
I think there is a potential for catastrophic backtracking here, with `\s*\s*` effectively if there are no block comments

Reply:
`(?<block_comment> /\*.*\*/ )` is not optional, so the whole `(?: \s* (?<block_comment> /\*.*\*/ ) )*` wouldn't match if no comment is available. Hence the last `\s*` is the only one to stay then.

Do I miss something?

Reply:
Nope, sorry - just me trying to review directly on GitHub (i.e. without syntax highlighting) while looking after the kids :sweat_smile: